### PR TITLE
[SPARK-47707][SQL] Special handling of JSON type for MySQL Connector/J 5.x

### PR DESCRIPTION
**Pull request by beryllw**
### What changes were proposed in this pull request?

MySQL JDBC driver `mysql-connector-java-5.1.49.jar` converts JSON type into Types.CHAR with a precision of Int.Max.
When receiving CHAR with Int.Max precision, Spark executor will throw an error of `java.lang.OutOfMemoryError: Requested array size exceeds VM limit `.

### Why are the changes needed?

before plan
```
+----------------------------------------------------+
|                        plan                        |
+----------------------------------------------------+
| == Physical Plan ==
*(1) Project [staticinvoke(class org.apache.spark.sql.catalyst.util.CharVarcharCodegenUtils, StringType, readSidePadding, field_test#75, 2147483647, true, false, true) AS field_test#97]
+- *(1) Scan org.apache.spark.sql.execution.datasources.v2.jdbc.JDBCScan$$anon$1@61874508 [field_test#75] PushedFilters: [], ReadSchema: struct<field_test:string>

 |
+----------------------------------------------------+
```
after this pr
```
+----------------------------------------------------+
|                        plan                        |
+----------------------------------------------------+
| == Physical Plan ==
*(1) Scan org.apache.spark.sql.execution.datasources.v2.jdbc.JDBCScan$$anon$1@32d7eba2 [field_test#28] PushedFilters: [], ReadSchema: struct<field_test:string>

 |
+----------------------------------------------------+
```

### Does this PR introduce _any_ user-facing change?

No

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by yaooqinn** [comment]
> Some MySQL JDBC drivers

The change itself looks safe to me, but can you clarify what kind of driver you were using. Is it standard or practical? Otherwise, it looks better to be customed by third-party developers.

**Comment by beryllw** [comment]
> The change itself looks safe to me, but can you clarify what kind of driver you were using. Is it standard or practical? Otherwise, it looks better to be customed by third-party developers.

mysql-connector-java-5.1.49.jar `com.mysql.jdbc.Driver`

**Comment by yaooqinn** [comment]
OK.

SPARK-43362 is already in a public release, we can create a new JIRA ticket. Maybe you can add a subtask for SPARK-47361

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala [hunk 1]
```diff
@@ -110,8 +110,9 @@ private case class MySQLDialect() extends JdbcDialect with SQLConfHelper {
       case Types.VARCHAR if "TINYTEXT".equalsIgnoreCase(typeName) =>
         // TINYTEXT is Types.VARCHAR(63) from mysql jdbc, but keep it AS-IS for historical reason
         Some(StringType)
-      case Types.VARCHAR if "JSON".equalsIgnoreCase(typeName) =>
-        // Some MySQL JDBC drivers converts JSON type into Types.VARCHAR with a precision of -1.
+      case Types.VARCHAR | Types.CHAR if "JSON".equalsIgnoreCase(typeName) =>
+        // Some MySQL JDBC drivers converts JSON type into Types.VARCHAR with a precision of -1
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala:114]: Please improve the comment here with 
- MySQL Connector/J 5.x as an example 
- the link https://github.com/mysql/mysql-connector-j/blob/release/5.1/src/com/mysql/jdbc/MysqlDefs.java#L295

**Review by yaooqinn — APPROVED** [review]
LGTM, except one minor comment

**Comment by yaooqinn** [comment]
Thank you @beryllw and welcome to Apache Spark Community,

 Merged to master

**Pull request closed without merging**